atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/remoting/iterators/RemoteRowColumnRangeIterator.java

```diff
@@ -0,0 +1,162 @@
+        if (keyValueService == null) {
+            throw new IllegalStateException(
+                    "This remote keyvalue service needs to be wrapped with RemotingKeyValueService.createClientSide!");
+        }
+
+        Cell lastCell = page.get(page.size() - 1).getKey();
+        byte[] row = lastCell.getRowName();
+        Preconditions.checkArgument(!RangeRequests.isLastRowName(lastCell.getColumnName()));
+        byte[] nextCol = RangeRequests.nextLexicographicName(lastCell.getColumnName());
+        RemoteRowColumnRangeIterator result = getMoreRows(keyValueService, tableRef, row, nextCol);
+        swapWithNewRows(validateIsRangeIterator(result));
+
+        if (position < page.size()) {
+            return page.get(position++);
+        } else {
+            throw new IllegalStateException();
+        }
+    }
+
+    protected RemoteRowColumnRangeIterator getMoreRows(KeyValueService kvs, TableReference tableRef, byte[] row, byte[] nextCol) {
+        ColumnRangeSelection newColumnRange = new ColumnRangeSelection(nextCol, columnRangeSelection.getEndCol(), columnRangeSelection.getBatchHint());
+        Map<byte[], RowColumnRangeIterator> result = kvs.getRowsColumnRange(tableRef, ImmutableList.of(row), newColumnRange, timestamp);
+        if (result.isEmpty()) {
+            new RemoteRowColumnRangeIterator(tableRef, columnRangeSelection, timestamp, false, ImmutableList.of());
+        }
+        RowColumnRangeIterator it = Iterables.getOnlyElement(result.values());
+        List<Map.Entry<Cell, Value>> page = ImmutableList.copyOf(Iterators.limit(it, columnRangeSelection.getBatchHint()));
+        return new RemoteRowColumnRangeIterator(tableRef, columnRangeSelection, timestamp, it.hasNext(), page);
+    }
+
+    private void swapWithNewRows(RemoteRowColumnRangeIterator other) {
+        hasNext = other.hasNext;
+        page = other.page;
+        position = 0;
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (position < page.size()) {
+            return true;
+        }
+        return hasNext;
+    }
+
+    @Override
+    public void remove() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        RemoteRowColumnRangeIterator that = (RemoteRowColumnRangeIterator) o;
+
+        if (timestamp != that.timestamp) return false;
+        if (hasNext != that.hasNext) return false;
+        if (position != that.position) return false;
+        if (tableRef != null ? !tableRef.equals(that.tableRef) : that.tableRef != null) return false;
+        if (columnRangeSelection != null ? !columnRangeSelection.equals(that.columnRangeSelection) : that.columnRangeSelection != null)
+            return false;
+        return page != null ? page.equals(that.page) : that.page == null;
```

Comment:
I don't really know - I just mirrored what RemotingKeyValueService does for RangeIterator since I couldn't find anyone who remembered why it was written that way.